src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -194,44 +199,69 @@ def register_options(cls, register):
            "`--setup-py2-args=\"bdist_wheel --python-tag py36.py37\"`. If unspecified, we just "
            "dump the setup.py chroot."
     )
+    register(
+      '--recursive', type=bool,
+      help="If specified, will run the setup.py command recursively on all exported targets that "
+           "the specified targets depend on, in dependency order.  This is useful, e.g., when "
+           "the command publishes dists, to ensure that any dependencies of a dist are published "
+           "before it."
+    )
 
 
 class SetupPy(Goal):
   subsystem_cls = SetupPyOptions
 
 
+def validate_args(args: Tuple[str, ...]):
+  # We rely on the dist dir being the default, so we know where to find the created dists.
+  if '--dist-dir' in args or '-d' in args:
+    raise InvalidSetupPyArgs('You must not set --dist-dir in your setup.py args')
+  # We don't allow publishing via setup.py, as we don't want the setup.py running rule,
+  # which is not a @console_rule, to side-effect (plus, we'd need to ensure that publishing
+  # happens in dependency order).
+  # TODO: A `publish` rule, that can invoke Twine to do the actual uploading.
+  if 'upload' in args:
```

Comment:
Also 'register', which reaches out and reserves a pypi project name. This may be noop now though, I'm not sure on deprecation status.

Reply:
Hmm turns out `upload` and `register` were removed in setuptools 42.0.0, but we should probably still check for them, in case someone downgrades setuptools in pants.ini.